### PR TITLE
Add simple Neo4j delta merge

### PR DESCRIPTION
**Pull request by pstutz**
Co-Authored-By: Max Kießling <max@kopfueber.org>

### CODE REVIEW

File: okapi-neo4j-io-testing/src/main/scala/org/opencypher/okapi/neo4j/io/testing/Neo4jServerFixture.scala [hunk 1]
```diff
@@ -45,16 +45,17 @@ trait Neo4jServerFixture extends BaseTestFixture {
...
 
   abstract override def beforeAll(): Unit = {
     super.beforeAll()
-    neo4jServer = TestServerBuilders
+    neo4jServer = EnterpriseTestServerBuilders
       .newInProcessBuilder()
-      .withConfig("dbms.security.auth_enabled", "true")
-      .withFixture(userFixture)
+      //.withConfig("dbms.security.auth_enabled", "true")
+      //.withFixture(userFixture)
```
Comment (Mats-SX) [inline okapi-neo4j-io-testing/src/main/scala/org/opencypher/okapi/neo4j/io/testing/Neo4jServerFixture.scala]: This certainly needs to be added back, no?

Comment (pstutz) [reply okapi-neo4j-io-testing/src/main/scala/org/opencypher/okapi/neo4j/io/testing/Neo4jServerFixture.scala]: It seems as if the tests are fine without it. It is surprising that it works without the user fixture though.

Comment (Mats-SX) [reply okapi-neo4j-io-testing/src/main/scala/org/opencypher/okapi/neo4j/io/testing/Neo4jServerFixture.scala]: Perhaps we're only making use of the `dataFixture` on the line below

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+      logger.info(s"Creating node key constraints: ${nodeKeyConstraints.mkString("\n\t- ", "\n\t- ", "")}")
+      session.run(nodeKeyConstraints.mkString("\n")).consume()
+
+      val idIndexes = maybeMetaLabel match {
+        case None =>
+          val commands = entityKeys.nodeKeys.keySet.flatten.map(label => s"CREATE INDEX ON :$label($metaPropertyKey)")
+          logger.info(s"Creating indexes for node entity keys:${commands.mkString("\n\t- ", "\n\t- ", "")}")
+          commands
+        case Some(ml) =>
+          val command = s"CREATE INDEX ON :$ml($metaPropertyKey)"
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: I wonder if we should create a unique constraint here instead?

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+          Set(command)
+      }
+
+      session.run(idIndexes.mkString("\n")).consume()
+    }
+  }
+
+  /**
+    * Merges the given `graph` into the sub-graph specified by `graphName` within an existing Neo4j database.
+    * Existing properties in the Neo4j graph are overwritten, missing ones are added.
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: Perhaps rephrase to something like

>Properties that exist in both the Neo4j graph and the argument graph will be updated in the Neo4j graph with the new value in the argument graph.

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    * Nodes and relationships are identified by using their entity keys. Therefore an entity key must be specified for
+    * every label combination present in the merge graph. Relationship keys are optional, if none are provided,
+    * then there can be at most one relationship with a given type between two nodes.
+    *
+    * @param graphName which sub-graph to sync the delta to
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: which subgraph `in the Neo4j database` to sync

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+  private[opencypher] def merge(
+    maybeGraphName: Option[GraphName],
+    graph: PropertyGraph,
+    config: Neo4jConfig,
+    entityKeys: EntityKeys
+  )(implicit caps: CAPSSession): Unit = {
+    val executorCount = caps.sparkSession.sparkContext.statusTracker.getExecutorInfos.length
+    implicit val executionContext: ExecutionContextExecutorService =
+      ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(executorCount))
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: We should replace this with the global pool

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    logger.debug(s"Using $executorCount Threads")
+
+    val maybeMetaLabel = maybeGraphName.map(gn => metaLabelForSubGraph(gn))
+
+    val writesCompleted = for {
+      _ <- Future.sequence(MergeWriters.writeNodes(maybeMetaLabel, graph, config, entityKeys.nodeKeys))
+      _ <- Future.sequence(MergeWriters.writeRelationships(maybeMetaLabel, graph, config, entityKeys.relKeys))
+      _ <- Future {
+        config.withSession { session =>
+          session.run(s"MATCH (n${maybeMetaLabel.getOrElse("")}) REMOVE n.$metaPropertyKey")
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: .consume()

File: spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        }
+      }
+    } yield Future {}
+    Await.result(writesCompleted, Duration.Inf)
```
Comment (Mats-SX) [inline spark-cypher/src/main/scala/org/opencypher/spark/api/io/neo4j/sync/Neo4jSync.scala]: Consider allowing timeout to be passed in

**Pull request merged**